proxy/src/connection.rs

```diff
@@ -254,6 +250,39 @@ impl Future for ConditionallyUpgradeServerToTls {
     }
 }
 
+impl ConditionallyUpgradeServerToTlsInner {
+    /// Polls the underlying socket for more data and buffers it.
+    ///
+    /// The buffer is matched for a TLS client hello message.
+    ///
+    /// `None` is returned if the underlying socket has closed.
+    fn poll_match_client_hello(&mut self) -> Poll<Option<tls::conditional_accept::Match>, io::Error> {
+        let sz = try_ready!(self.socket.read_buf(&mut self.peek_buf));
+        if sz == 0 {
+            return Ok(None.into())
+        }
+
+        let buf = self.peek_buf.as_ref();
+        Ok(Some(tls::conditional_accept::match_client_hello(buf, &self.tls.identity)).into())
+    }
+}
+
+impl From<ConditionallyUpgradeServerToTlsInner> for tls::UpgradeServerToTls {
+    fn from(inner: ConditionallyUpgradeServerToTlsInner) -> Self {
+        tls::Connection::accept(inner.socket, inner.peek_buf.freeze(), inner.tls.config)
+    }
+}
+
+impl From<ConditionallyUpgradeServerToTlsInner> for Connection {
+    fn from(inner: ConditionallyUpgradeServerToTlsInner) -> Self {
+        Connection::plain_with_peek_buf(
```

Comment:
Instead of making this a `From` implementation, just change the signature of `plain_with_peek_buf` to take a `ConditionallyUpgradeServerToTlsInner` parameter. In particular, `Connection::from(x)` is not explicit enough that we're creating a *plaintext*, i.e. non-secure, connection.

Reply:
Instead, I added more explicitly named helpers onto `ConditionallyUpgradeServerToTlsInner` (so that the details of the type are self-contained). Let me know what you think.